Source/JavaScriptCore/runtime/JSStringIterator.cpp

```diff
@@ -59,3 +59,50 @@ JSStringIterator* JSStringIterator::clone(ExecState* exec)
 }
 
 } // namespace JSC
+
+#if PLATFORM(IOS) && !USE(APPLE_INTERNAL_SDK)
+U_STABLE int32_t U_EXPORT2
+ios_specific_unorm_normalize(const UChar *source, int32_t sourceLength,
+                             UNormalizationMode mode, int32_t options,
+                             UChar *resultBuffer, int32_t resultLength,
+                             UErrorCode *status) {
+    if (U_FAILURE(*status)) {
+        return 0;
+    }
+    
+    if ( (source == NULL ? sourceLength != 0 : sourceLength < -1) ||
+        (resultBuffer == NULL ? resultLength != 0 : resultLength < 0) ||
+        (source == resultBuffer && source != NULL)
+        ) {
+        *status = U_ILLEGAL_ARGUMENT_ERROR;
+        return 0;
+    }
+    
+    RetainPtr<CFMutableStringRef> normalizedString = adoptCF(CFStringCreateMutable(kCFAllocatorDefault, 0));
+    CFStringAppendCharacters(normalizedString.get(), source, sourceLength);
+    CFStringNormalizationForm normalizationForm;
+    switch (mode) {
+        case UNORM_NFC:
+            normalizationForm = kCFStringNormalizationFormC;
+            break;
+        case UNORM_NFD:
+            normalizationForm = kCFStringNormalizationFormD;
+            break;
+        case UNORM_NFKC:
+            normalizationForm = kCFStringNormalizationFormKC;
+            break;
+        case UNORM_NFKD:
+            normalizationForm = kCFStringNormalizationFormKD;
+            break;
+        default:
+            RELEASE_ASSERT_NOT_REACHED();
+            break;
+    }
+    
+    CFStringNormalize(normalizedString.get(), normalizationForm);
+    if (resultLength > 0 && resultBuffer != nullptr) {
+        CFStringGetCharacters(normalizedString.get(), CFRangeMake(0, resultLength), resultBuffer);
```

Comment:
According to the [header docs](http://opensource.apple.com//source/WebCore/WebCore-514/icu/unicode/unorm.h) in the case where the result length is less than the normalizedString length you should set the error code to ```U_BUFFER_OVERFLOW_ERROR```.